docs/design-proposals/apiserver/openstack-api-support.md

```diff
@@ -0,0 +1,408 @@
+
+|Name	    |In	  |Type	|Description|
+|-----------|-----|-------|-----------|
+|server_id	|path	|string	|The UUID of the server.|
+
+Response
+
+There is no body content for the response of a successful DELETE query
+
+
+### VM Actions
+The /servers/{server-id}/action will take an action defined in the request, the below uses reboot VM as an example.
+
+Request
+
+|Name	|In	  |Type	|Description|
+|------|----|-------|-----------|
+|server_id	|path	|string	|The UUID of the server.|
+|reboot	|body	|object	|The action to reboot a server.|
+|type	|body	|string	|The type of the reboot action. The valid values are HARD and SOFT. A SOFT reboot attempts a graceful shutdown and restart of the server. A HARD reboot attempts a forced shutdown and restart of the server. The HARD reboot corresponds to the power cycles of the server.|
+
+Example Reboot Server (reboot Action)
+```
+{
+    "reboot" : {
+        "type" : "HARD"
+    }
+}
+```
+Response¶
+If successful, this method does not return content in the response body.
+
+## Implementation
+List of changes in Arktos:
+1.	New routes registered to the API server, for servers and actions
+2.	New set of handlers for each route for VM and actions ( Reuse of existing pod handlers )
+3.	Converting logic to convert Openstack requests to Arktos request and Response to Openstack response in the web service, for both Server and Action requests and responses
+4.	API server logic to retrieve the tenant info from Openstack request header, to construct the desired PATH for the VM request to Arktos POD (or ACTION) routes.
+5.	(Stretch goal) modify Openstack client SDK to for Arktos VM APIs
+
+## API Security 
+
+API authentication and authorization, and audit chain are built into the Arktos API server already. The new route will be part of the chain and leverage the security control there as the other objects such as PODs, actions built into Arktos.
+
+
+## Support Openstack testing tool with minimum modification from users ( Goal #3)
+As indicated in the appendix section, Openstack client components make interactions with multiple Openstack services, including neutron, compute and keystone services for network, image, flavor, authentication purposes. In order to have similar user experience when running those clients against Arktos, some extra routes will be added on:
+
+	{Arktos-service-url}/images
+	{Arktos-service-url}/flavors
+
+### Handel Namespace
+TBD -- ideally, the namespace in Arktos should map to Project in Openstack. for this release, it will be simply using the 
+default namespace for tenant's VM workloads.
+
+### VM flavor support:
+TBD: Arktos to have a config map to preserve basic Openstack flavor
+
+### VM imageRef support
+TBD: Eventually there will be an image registry service to support this. For this current design, instead of a full flether image registry service, Arktos can have a config map to perserve commom images such as coros and Ubuntu etc.
+
+## Future works
+### Create VMs in batch
```

Comment:
Goal

Reply:
fixed as mentioned for VM batch creation optimization.